### DISCUSSION TITLE
no column found for name: A_

### DISCUSSION THREAD

**Discussion by brsnik** [SeaQL/sea-orm#2128] [Q&A]
### Discussed in https://github.com/SeaQL/sea-orm/discussions/2124

<div type='discussions-op-text'>

<sup>Originally posted by **brsnik** March  1, 2024</sup>
This builds the correct query in terms of SQL (from what I can see in the tracer).
However I get error `Query Error: no column found for name: A_barcode`

```
MvB::find()
        .find_with_related(MvA)
//        .filter(MvBColumn::Barcode.eq(gtin))
        .stream_partial_model::<DatabaseConnection, MvBSearchModel, MvASearchModel>(&conn)
        .await
```

My partial models:
```
#[derive(Debug, FromQueryResult, DerivePartialModel)]
#[sea_orm(entity = "Entity")]
pub struct MvBSearchModel {
    pub barcode: String,
    pub lbl_type: String,
}
```
and 
```
#[derive(Debug, FromQueryResult, DerivePartialModel)]
#[sea_orm(entity = "Entity")]
pub struct MvASearchModel {
    pub prod_art_id: i32,
    pub prod_art_pub_uuid: Uuid,
    pub lbl_status: String,
}
```

The query that gets built, again it's correct:
```
SELECT 
    "mv_b"."barcode", 
    "mv_b"."lbl_type", 
    "mv_a"."prod_art_id", 
    "mv_a"."prod_art_pub_uuid", 
    "mv_a"."lbl_status" 
FROM "products"."mv_b" 
    LEFT JOIN "products"."mv_a" ON "mv_b"."prod_art_pvt_id" = "mv_a"."prod_art_pvt_id" 
ORDER BY "mv_b"."prod_brcd_id" ASC
```

What is causing this error? Am I doing something wrong?
</div>

**Comment by tyt2y3** [comment]
What are you trying to do? Why do you think it's incorrect, and what is the expected behaviour?

**Reply by brsnik** [reply]
The expected behaviour is I need to get partial data (aka some columns) from both MvB and MvA (A left joined to B) as a stream. I've not even been able to get it to work without a stream also, by using `.into_partial_model`.
It works fine if I do `.into_partial_model::<MvBSearchModel>()` with only MvBSearchModel and only returns the requested data. But the moment I add both `.into_partial_model::<MvBSearchModel, MvASearchModel>()` with `.find_also_related(MvA)` it throws the same error mentioned above `Query Error: no column found for name: A_barcode`.

I'm not sure if I'm doing something wrong, or it has to do with a bug of some sorts. Please let me know @tyt2y3?

So when I do this, all columns are selected from MvB & MvA.
```
let x = MvB::find()
    .find_also_related(MvA)
    .stream(&conn)
    .await;
```
But the result I get is just for data from MvB and nothing from MvA: 
```
(Model { prod_brcd_id: 1, prod_art_pvt_id: 2, t_latest: 2024-02-27T11:49:48.817816, barcode: "52087", prod_brcd_ty_id: 1, lbl_type: "gtin", sts_deleted: false }, None)
```

When I try to use partials, the query that is build is correct and it only selects the columns I need. But I get the error mentioned above `Query Error: no column found for name: A_barcode`. I'm honestly not sure what `A_barcode` means. It's not a table name of mine, and as you can see in MvB the column name is just `barcode`. I'm guessing internally it's mapping things a certain way. But I have no clue if I'm doing something wrong, or it's a bug of sorts.
```
let x = MvB::find()
    .find_also_related(MvA)
    .stream_partial_model::<DatabaseConnection, MvBSearchModel, MvArSearchModel>(&conn)
    .await;
```

And if I use `.select_only()` and specify the columns, I get an error that columns are missing.

The MvB Model and Relations for reference.
```
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(schema_name = "products", table_name = "mv_b")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub prod_brcd_id: i32,
    pub prod_art_pvt_id: i32,
    pub t_latest: DateTime,
    pub barcode: String,
    pub prod_brcd_ty_id: i32,
    pub lbl_type: String,
    pub sts_deleted: bool,
}

#[derive(Debug, FromQueryResult, DerivePartialModel)]
#[sea_orm(entity = "Entity")]
pub struct MvBSearchModel {
    pub barcode: String,
    pub lbl_type: String,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::mv_a::Entity",
        from = "Column::ProdArtPvtId",
        to = "super::mv_a::Column::ProdArtPvtId",
        on_update = "NoAction",
        on_delete = "NoAction"
    )]
    MvA,
}

impl Related<super::mv_a::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::MvA.def()
    }
}
```

**Reply by brsnik** [reply]
@tyt2y3 Can I please get some feedback?

**Comment by Raz-Hemo** [comment]
I also ran into this issue, this is a real bummer. I don't think there's any good way of using both find_with_related and a partial model.

**Comment by zefr0x** [comment]
I believe this is a current limitation in sea-orm.

When using `.find_with_related()`, it will assign aliases for column names using the `A_` prefix for the first table and `B_` prefix for the second. The problem is that after using `.into_partial_model()` it will reset those aliases, so when sea-orm later parses the returned values  it will not see `A_barcode` is your case. 

The same thing when using `.select_only()` it will reset the selects and after using `.column(...)` or `.columns(...)` it will not write those aliases again.


The current workaround is to manually use `.column_as(..., ...)` to write the aliases yourself, in your case it should be something like that:
```rust
MvB::find()
        .find_with_related(MvA)
        .select_only()
        .column_as(mvb::Column::Barcode, "A_barcode")
        .column_as(mvb::Column::LblType, "A_lbl_type")
        .column_as(mva::Column::ProdArtId, "B_prod_art_id")
        .column_as(mva::Column::ProdArtPubUuid, "B_prod_art_pub_uuid")
        .column_as(mva::Column::LblStatus, "B_lbl_status")
        .into_model::<DatabaseConnection, MvBSearchModel, MvASearchModel>()
        .await
```

And the resulting query should look like:
```sql
SELECT 
    "mv_b"."barcode" as A_barcode, 
    "mv_b"."lbl_type" as A_lbl_type, 
    "mv_a"."prod_art_id" as B_prod_art_id, 
    "mv_a"."prod_art_pub_uuid" as B_prod_art_pub_uuid, 
    "mv_a"."lbl_status" as B_lbl_status 
FROM "products"."mv_b" 
    LEFT JOIN "products"."mv_a" ON "mv_b"."prod_art_pvt_id" = "mv_a"."prod_art_pvt_id" 
ORDER BY "mv_b"."prod_brcd_id" ASC
```
